### PR TITLE
Skip grpc-accept-encoding when the only option is identity

### PR DESCRIPTION
**Pull request by tkountis**
Motivation:
`grpc-accept-encoding` and `grpc-encoding` headers can be skipped when the only value associated with is `identity`. The spec allows this relaxation and we can save processing and wire time.

Modifications:
Skip the relevant headers when the value is only `identity`.

Result:
Less headers and less processing done to handle unnecessary data.

### CODE REVIEW

File: servicetalk-encoding-api-internal/src/main/java/io/servicetalk/encoding/api/internal/HeaderUtils.java [hunk 1]
```diff
@@ -43,13 +43,13 @@ private HeaderUtils() {
      * <p>
      * If no supported encodings are configured then the result is always {@code null}
      * If no accepted encodings are present in the request then the result is always {@code null}
-     * In all other cases, the first matching encoding (that is NOT {@link ContentCodings#identity()}) is preferred,
```
Comment (bondolo) [inline servicetalk-encoding-api-internal/src/main/java/io/servicetalk/encoding/api/internal/HeaderUtils.java:46]: Worth deprecating ContentCodings#identity() ?

Comment (tkountis) [reply servicetalk-encoding-api-internal/src/main/java/io/servicetalk/encoding/api/internal/HeaderUtils.java:46]: The class `ContentCodings` is deprecated.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java [hunk 2]
```diff
@@ -284,9 +285,10 @@ private static void initResponse(final HttpResponseMetaData response, @Nullable
         final HttpHeaders headers = response.headers();
         headers.set(SERVER, GRPC_USER_AGENT);
         headers.set(CONTENT_TYPE, GRPC_CONTENT_TYPE);
-        if (context != null) {
-            headers.set(GRPC_ACCEPT_ENCODING_KEY,
-                    acceptedEncodingsHeaderValueOrCached(context.supportedMessageCodings()));
+        final CharSequence acceptedEncoding = context == null ? null :
```
Comment (Scottmitch) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java:288]: consider avoiding a `null` conditional if not necessary:
```suggestion
          if (context != null) {
            final CharSequence acceptedEncoding =
                    acceptedEncodingsHeaderValueOrCached(context.supportedMessageCodings());
            if (acceptedEncoding != null) {
                headers.set(GRPC_ACCEPT_ENCODING_KEY, acceptedEncoding);
            }
        }
```

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java [hunk 3]
```diff
@@ -347,7 +352,7 @@ private static CharSequence acceptedEncodingsHeaderValue0(final List<ContentCode
             builder.append(codec.name());
         }
 
-        return newAsciiString(builder.toString());
+        return builder.length() > 0 ? newAsciiString(builder.toString()) : null;
```
Comment (Scottmitch) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java:355]: suggestions:
- avoid conditional inside loop
- pre-size the string builder to reduce chances of copy/resize

```java
private static final String CONTENT_ENCODING_SEPARATOR = ", ";
```
```suggestion
        StringBuilder builder = new StringBuilder(codings.size() * (12 + CONTENT_ENCODING_SEPARATOR.length()));
        for (ContentCodec codec : codings) {
            if (codec == identity()) {
                continue;
            }

            builder.append(codec.name()).append(CONTENT_ENCODING_SEPARATOR);
        }
        if (builder.length() > CONTENT_ENCODING_SEPARATOR.length()) {
            builder.setLength(builder.length() - CONTENT_ENCODING_SEPARATOR.length());
            return newAsciiString(builder.toString());
        }
        return null;
```

File: servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/GrpcMessageEncodingTest.java [hunk 4]
```diff
@@ -421,16 +413,19 @@ private TesterClient newClient(@Nullable final List<ContentCodec> supportedCodin
...
+            throwAsyncErrors();
+            throw t;
         }
-
-        verifyNoErrors();
     }
 
-    private void verifyNoErrors() throws Throwable {
+    private void throwAsyncErrors() throws Throwable {
         if (!errors.isEmpty()) {
```
Comment (Scottmitch) [inline servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/GrpcMessageEncodingTest.java:429]: nit: make this a concurrent queue to replace `empty` / `get(0)` (sensitive to race conditions) with `Object o = queue.poll()`

File: servicetalk-encoding-api-internal/src/main/java/io/servicetalk/encoding/api/internal/HeaderUtils.java (hunk 1 above)
Comment (Scottmitch) [reply servicetalk-encoding-api-internal/src/main/java/io/servicetalk/encoding/api/internal/HeaderUtils.java:46]: does this PR remove the last usage of the class?

**Review by Scottmitch — APPROVED** [review]
few comments/suggestions then lgtm

File: servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/GrpcMessageEncodingTest.java (hunk 4 above)
Comment (tkountis) [reply servicetalk-grpc-netty/src/test/java/io/servicetalk/grpc/netty/GrpcMessageEncodingTest.java:429]: Changed to keep consistent with other test classes. FWIW, there is no race here, the List was `synchronized` and the `throw` is called after the Futures are resolved.

**Comment by tkountis** [comment]
> does this PR remove the last usage of the class?
@Scottmitch yes, there were a few places that weren't caught in the previous refactoring. I fixed them here.

File: servicetalk-encoding-api-internal/src/main/java/io/servicetalk/encoding/api/internal/HeaderUtils.java (hunk 1 above)
Comment (tkountis) [reply servicetalk-encoding-api-internal/src/main/java/io/servicetalk/encoding/api/internal/HeaderUtils.java:46]: yes, there were a few places that weren't caught in the previous refactoring. I fixed them here.

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java [hunk 5]
```diff
@@ -324,30 +329,33 @@ private static Status getStatusDetails(final HttpHeaders headers) {
...
     private static CharSequence acceptedEncodingsHeaderValueOrCached(final List<ContentCodec> codings) {
         return ENCODINGS_HEADER_CACHE.computeIfAbsent(codings, (__) -> acceptedEncodingsHeaderValue0(codings));
     }
 
+    @Nullable
     private static CharSequence acceptedEncodingsHeaderValue0(final List<ContentCodec> codings) {
-        StringBuilder builder = new StringBuilder();
+        StringBuilder builder = new StringBuilder(codings.size() * (12 + CONTENT_ENCODING_SEPARATOR.length()));
         for (ContentCodec codec : codings) {
             if (codec == identity()) {
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java:347]: Users may have their own version of "identity" codec. We should implement `equals` and `hashCode` for `IdentityContentCodec` then use `identity().equals(codec)` here and in `HeaderUtils` for HTTP.

Comment (tkountis) [reply servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java:347]: Agreed, issue: https://github.com/apple/servicetalk/issues/1455

**Pull request merged**